feat(i18n): add backwards compatible async messages

montezume opened the pull request:
#### Adds backwards compatible option for async loading of applicationMessages.

#### Summary

When <AsyncLocaleData> is given the prop `fetchApplicationMessages`, of type function, and that function is invoked with a language, which results in a promise, then `AsyncLocaleData` will resolve that promise and invoke `children` with its result. Else, it will just invoke children with the property `applicationMessages`, which is passed into it from app-shell.

montezume commented on packages/application-shell/src/components/application-shell/application-shell.js:
```diff
@@ -345,7 +358,8 @@ RestrictedApplication.propTypes = {
   render: PropTypes.func.isRequired,
-  applicationMessages: PropTypes.object.isRequired,
+  applicationMessages: PropTypes.object,
+  fetchApplicationMessages: PropTypes.func,
```
maybe oneOf or something? I'm not sure how to make a propType be one or the other

tdeekens replied:
You can do

```js
applicationMessages: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.instanceOf(Whatever)
])
```

or any other type combination.

montezume replied:
yeah that makes sense 👍

montezume replied:
sort of weird though because applicationMessages needs to be a function, that receives a language, and returns a promise 🤔

tdeekens replied:
Yeah it's a `applicationMessages(language: Language) => Promise<ApplicationMessage>` as far as I understand.

montezume commented on packages/i18n/async-locale-data/async-locale-data.js:
```diff
@@ -35,12 +42,28 @@ class AsyncLocaleData extends React.Component {
           language,
-          messages: data,
+          messages: response[0],
+          applicationMessages: response[1]
```
if no promise result, then just use the passed in applicationMessages

tdeekens commented on packages/i18n/async-locale-data/async-locale-data.js:
```diff
@@ -35,12 +42,28 @@ class AsyncLocaleData extends React.Component {
+      promises = promises.concat([applicationMessagePromise]);
+    }
+
+    Promise.all(promises).then(
```
Maybe it eases the eye if we'd do a 

```js
const messageFetchingPromises = do {
  if(isPromiseLike(applicationMessagePromise)) 
    [loadI18n(language), applicationMessagePromise];
  else
    [loadI18n(language)]
}
const [messages, applicationMessages] = await Promise.all(messageFetchingPromises);

this.setState({...})
```

montezume commented on packages/i18n/async-locale-data/async-locale-data.js:
```diff
@@ -1,11 +1,15 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { reportErrorToSentry } from '@commercetools-frontend/sentry';
+import isFunction from 'lodash.isfunction';
```
removed it 👍

montezume commented on packages/application-shell/README.md:
```diff
@@ -65,13 +65,13 @@ ReactDOM.render(<EntryPoint />, document.getElementById('root'));
-| `onRegisterErrorListeners` | `func`   |    ✅    | -       | A callback function to setup global event listeners, called when the `ApplicationShell` is mounted                                            |
+| Props                      | Type               | Required | Default | Description                                                                                                                                                                                                       |
+| -------------------------- | ------------------ | :------: | ------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `applicationMessages`      | `object` or `func` |    ✅    | -       | Either an object containing all the translated messages per locale (`{ "en": { "Welcome": "Welcome" }, "de": { "Welcome": "Wilkommen" }}`), or a function that returns a promise that resolves to such an object. |
```
this was outdated

emmenko replied:
I would document how this works. Maybe add a section below about `## Loading i18n messages with code splitting`

montezume commented on packages/i18n/package.json:
```diff
@@ -34,6 +34,7 @@
     "build:es:watch": "npm run build:es -- -w"
   },
   "dependencies": {
+    "@babel/runtime": "7.1.5",
```
if you use an async function without having this as a dependency, then rollup will complain. we don't bundle dependencies with rollup anyways, so this won't increase bundle size

montezume commented:
Okay I addressed the PR comments and tried to update the documentation in `ApplicationShell`. I don't think we should add documentation for `AsyncLocaleData` itself, as it's not used externally - it's only used inside of AppShell, not inside of the apps directly.

emmenko commented on packages/i18n/async-locale-data/async-locale-data.js:
```diff
@@ -33,18 +39,38 @@ class AsyncLocaleData extends React.Component {
+        language,
+        messages: mergeMessages(
+          messages,
+          applicationMessages || this.props.applicationMessages[language]
```
Use `this.props.applicationMessages && this.props.applicationMessages[language]` since the prop is optional and could run into runtime errors

montezume replied:
👍

montezume replied:
applicationMessages is defined as being a required prop in the docs though, and the current propType is object.isRequired? Should we change it to optional?

tdeekens commented on playground/src/components/entry-point/entry-point.js:
```diff
@@ -11,6 +11,19 @@ import {
+
+const fetchApplicationMessages = lang =>
+  new Promise((resolve, reject) =>
+    import(`../../i18n/data/${lang}.json` /* webpackChunkName: "application-messages-[request]" */).then(
```
Insane.

montezume replied:
🤷‍♂️. We had this [request] thing before we did the split out anyways.

montezume commented:
@emmenko documented [here](https://github.com/commercetools/merchant-center-application-kit/pull/116/commits/9f05fabe6667b49dd20d2b002c8c897730c6e8d3)

emmenko commented on packages/application-shell/README.md:
````diff
@@ -63,6 +63,34 @@ EntryPoint.displayName = 'EntryPoint';
+
+```js
+// define a function that accepts a language, and returns a promise.
+const applicationMessages = lang =>
````
Nitpick: we can maybe name this variable `loadApplicationMessagesForLanguage`, to make it clear what it does

montezume commented:
anything from your side @tdeekens ?
